### PR TITLE
feat: add plugin-i18next

### PR DESCRIPTION
**Pull request by QuantumlyTangled**

### CODE REVIEW

File: packages/in17n/package.json [hunk 1]
```diff
@@ -0,0 +1,46 @@
+{
+	"name": "@sapphire/plugin-in17n",
+	"version": "1.0.0",
+	"description": "> TODO: description",
+	"author": "QuantumlyTangled <yo@quantumlytangled.com>",
```
Comment (Soumil07) [inline packages/in17n/package.json:5]:
```suggestion
	"author": "@sapphire",
```

**Review by Soumil07 — CHANGES_REQUESTED** [review]

File: packages/in17n/LICENSE.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# BSD 3-Clause License
```
Comment (kyranet) [inline packages/in17n/LICENSE.md:1]: Why not MIT?

File: packages/in17n/package.json [hunk 3]
```diff
@@ -0,0 +1,72 @@
+{
+	"name": "@sapphire/plugin-in17n",
+	"version": "1.0.0",
+	"description": "> TODO: description",
```
Comment (kyranet) [inline packages/in17n/package.json:4]: Needs description

File: packages/in17n/package.json [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+	"author": "@sapphire",
+	"maintainers": [
+		"QuantumlyTangled <yo@quantumlytangled.com>"
+	],
+	"contributors": [
+		"QuantumlyTangled <yo@quantumlytangled.com>"
+	],
+	"license": "BSD-3-Clause",
+	"main": "dist/index.js",
```
Comment (kyranet) [inline packages/in17n/package.json:13]: Please base the package.json from the other plugins, this is missing a bunch of stuff, including ESM support.

File: packages/in17n/src/register.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+		 *   const guild = await driver.getRepository(GuildEntity).findOne({ id: message.guild.id });
+		 *   return guild?.language ?? 'en-US';
+		 * };
+		 * ```
+		 */
+		fetchLanguage: (message: Message) => Awaited<string | null>;
+	}
+
+	export interface ClientOptions {
+		in17n?: In17nOptions;
```
Comment (kyranet) [inline packages/in17n/src/register.ts:57]: Missing docs

File: packages/in17n/src/lib/types/options.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+ * @since 1.0.0
+ */
+export interface In17nOptions {
+	/**
+	 * Used as the default 2nd to last fallback locale if no other is found.
+	 * It's only followed by "en-US".
+	 * @since 1.0.0
+	 */
+	defaultName?: string;
+	/**
```
Comment (kyranet) [inline packages/in17n/src/lib/types/options.ts:15]:
```suggestion
	defaultName?: string;

	/**
```

File: packages/in17n/src/lib/types/options.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+	 * The options passed to `backend` in `i18next.init`.
+	 * @since 1.0.0
+	 */
+	backend?: i18nextFsBackend.i18nextFsBackendOptions;
+	/**
```
Comment (kyranet) [inline packages/in17n/src/lib/types/options.ts:20]:
```suggestion
	backend?: i18nextFsBackend.i18nextFsBackendOptions;

	/**
```

File: packages/in17n/src/lib/types/options.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+	 * The options passed to `i18next.init`.
+	 * @since 1.0.0
+	 */
+	i18next?: InitOptions;
+	/**
```
Comment (kyranet) [inline packages/in17n/src/lib/types/options.ts:25]:
```suggestion
	i18next?: InitOptions;

	/**
```

File: packages/in17n/src/lib/types/options.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+	 * The directory in which "i18next-fs-backend" should search for files.
+	 * Defaulted to "<rootDirectory>/languages".
+	 * @since 1.0.0
+	 */
+	defaultLanguageDirectory?: string;
+	/**
```
Comment (kyranet) [inline packages/in17n/src/lib/types/options.ts:31]:
```suggestion
	defaultLanguageDirectory?: string;

	/**
```

File: packages/in17n/src/lib/types/options.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+	 * The default value to be used if a specific language key isnt found.
+	 * Defaulted to "default:default".
+	 * @since 1.0.0
+	 */
+	defaultMissingKey?: string;
+	/**
```
Comment (kyranet) [inline packages/in17n/src/lib/types/options.ts:37]:
```suggestion
	defaultMissingKey?: string;

	/**
```

File: packages/in17n/src/register.ts [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+		this.fetchLanguage = options.fetchLanguage ?? (() => options.in17n?.defaultName ?? null) ?? (() => 'en-US');
+	}
+
+	public static async [preLogin](this: SapphireClient): Promise<void> {
+		await this.in17n.init();
+	}
+}
+
+declare module 'discord.js' {
```
Comment (kyranet) [inline packages/in17n/src/register.ts:20]: This needs to be moved elsewhere, TypeScript doesn't pick augmentations in `/register`. Check the changes in the other two plugins to follow consistency.

File: packages/in17n/src/lib/In17nHandler.ts [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+	 * Describes whether `In17nHandler#init` has been run and languages are loaded in `In17nHandler.languages`.
+	 * @since 1.0.0
+	 */
+	public languagesLoaded = false;
+
+	/**
+	 * A `Map` of `i18next` language functions keyed by their language code.
+	 * @since 1.0.0
+	 */
+	public languages!: Map<string, TFunction>;
```
Comment (kyranet) [inline packages/in17n/src/lib/In17nHandler.ts:24]:
```suggestion
	public readonly languages = new Map<string, TFunction>();
```

File: packages/in17n/src/lib/In17nHandler.ts [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+					load: 'all',
+					defaultNS: this.options?.defaultNS ?? 'default',
+					ns: namespaces,
+					preload: languages
+				} as InitOptions,
+				this.options?.i18next
+			)
+		);
+
+		this.languages = new Map(languages.map((item) => [item, i18next.getFixedT(item)]));
```
Comment (kyranet) [inline packages/in17n/src/lib/In17nHandler.ts:75]:
```suggestion
		for (const item of languages) {
			this.languages.set(item, i18next.getFixedT(item));
		}
```

File: packages/in17n/src/lib/In17nHandler.ts [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+						escapeValue: false
+					},
+					load: 'all',
+					defaultNS: this.options?.defaultNS ?? 'default',
+					ns: namespaces,
+					preload: languages
+				} as InitOptions,
+				this.options?.i18next
+			)
+		);
```
Comment (kyranet) [inline packages/in17n/src/lib/In17nHandler.ts:73]: I have never been a fan of mutating options, and we don't use `mergeDefault` anywhere in Sapphire, not even plugins do.
```suggestion
		await i18next.init({
			backend: this.backendOptions,
			fallbackLng: this.options?.defaultName ?? 'en-US',
			initImmediate: false,
			interpolation: {
				escapeValue: false
			},
			load: 'all',
			defaultNS: this.options?.defaultNS ?? 'default',
			ns: namespaces,
			preload: languages,
			...(this.options?.i18next ?? {})
		});
```

File: packages/in17n/src/lib/In17nHandler.ts [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+		return language(
+			key,
+			mergeDefault(
+				{
+					defaultValue: language(this.options?.defaultMissingKey ?? 'default:default', { replace: { key } }),
+					replace
+				},
+				options
+			)
+		);
```
Comment (kyranet) [inline packages/in17n/src/lib/In17nHandler.ts:116]:
```suggestion
		return language(key, {
			defaultValue: language(this.options?.defaultMissingKey ?? 'default:default', { replace: { key } }),
			replace,
			...(options ?? {})
		});
```

File: packages/in17n/src/lib/In17nHandler.ts [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+	/**
+	 * @description Skips any files that don't end with `.json`.
+	 * @param dir The directory that should be walked.
+	 * @param namespaces The currently known namespaces.
+	 * @param folderName The currently walked folder.
+	 * @since 1.0.0
+	 * @protected
+	 */
+	protected async walkLanguageDirectory(dir: string, namespaces: string[] = [], folderName = '') {
+		const files = await readdir(dir);
```
Comment (kyranet) [inline packages/in17n/src/lib/In17nHandler.ts:128]: Use `fs/promises.opendir`, it's more convenient, uses less resources, and returns `Dirent`s, which can tell you whether something is a file, a directory, or any other thing, without extra calls.

Also uses for-await-of, which is pretty convenient as it allows for zero-copy operation.

File: packages/in17n/src/lib/In17nHandler.ts [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+	 * @param locale The language to be used.
+	 * @since 1.0.0
+	 */
+	public fetchT(locale: string) {
+		if (!this.languagesLoaded) throw new Error('Cannot call this method until In17nHandler#init has been called');
+
+		const t = this.languages.get(locale);
+		if (!t) throw new ReferenceError('Invalid language provided');
+
+		return t;
```
Comment (kyranet) [inline packages/in17n/src/lib/In17nHandler.ts:90]: Avoid negative conditions when they're one-liners:
```suggestion
		if (t) return t;
		throw new ReferenceError('Invalid language provided');
```

**Review by kyranet — CHANGES_REQUESTED** [review]

File: packages/in17n/package.json [hunk 18]
```diff
@@ -0,0 +1,72 @@
+{
+	"name": "@sapphire/plugin-in17n",
```
Comment (Favna) [inline packages/in17n/package.json:2]:
```suggestion
	"name": "@sapphire/plugin-i18next",
```

And reflect that change everywhere else as well

File: packages/in17n/LICENSE.md (hunk 2 above)
Comment (Favna) [reply packages/in17n/LICENSE.md:1]: This. Please make it MIT. In fact, please delete the entire file. Lerna is going to overwrite it with the LICENSE.md in root anyway.

File: packages/in17n/package.json [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+	"version": "1.0.0",
+	"description": "> TODO: description",
+	"author": "@sapphire",
+	"maintainers": [
+		"QuantumlyTangled <yo@quantumlytangled.com>"
+	],
+	"contributors": [
+		"QuantumlyTangled <yo@quantumlytangled.com>"
+	],
```
Comment (Favna) [inline packages/in17n/package.json:11]:
```suggestion
```

We use all-contributors and codeowners, not package.json fields.

File: packages/in17n/package.json [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+	"license": "BSD-3-Clause",
```
Comment (Favna) [inline packages/in17n/package.json:12]:
```suggestion
	"license": "MIT",
```

See above

File: packages/in17n/package.json [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+		"esm:default": "gen-esm-wrapper dist/index.js dist/index.mjs"
+	},
+	"dependencies": {
+		"@sapphire/utilities": "^1.2.0",
+		"i18next": "^19.8.4",
+		"i18next-fs-backend": "^1.0.7"
+	},
+	"peerDependencies": {
+		"@sapphire/framework": "1.0.0-alpha.0",
+		"@sapphire/pieces": "1.0.0",
```
Comment (Favna) [inline packages/in17n/package.json:28]:
```suggestion
		"@sapphire/framework": "1.0.0-alpha.1",
		"@sapphire/pieces": "1.0.1",
```

File: packages/in17n/package.json [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+		"@types/i18next-fs-backend": "^1.0.0",
+		"discord.js": "^12.5.1"
```
Comment (Favna) [inline packages/in17n/package.json:30]:
```suggestion
		"discord.js": "^12.x"
```

File: packages/in17n/package.json [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+	},
+	"dependencies": {
+		"@sapphire/utilities": "^1.2.0",
+		"i18next": "^19.8.4",
+		"i18next-fs-backend": "^1.0.7"
+	},
+	"peerDependencies": {
+		"@sapphire/framework": "1.0.0-alpha.0",
+		"@sapphire/pieces": "1.0.0",
+		"@types/i18next-fs-backend": "^1.0.0",
```
Comment (Favna) [inline packages/in17n/package.json:29]:
```suggestion
		"@types/i18next-fs-backend": "^1.x",
```

File: packages/in17n/package.json [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+	},
+	"keywords": [
+		"sapphire-project",
+		"plugin",
+		"bot",
+		"typescript",
+		"ts",
+		"yarn",
+		"discord",
+		"sapphire"
```
Comment (Favna) [inline packages/in17n/package.json:64]:
```suggestion
		"discord",
		"sapphire",
		"i18next",
		"i18n"
```

SEO goes brrrr

File: packages/in17n/src/register.ts (hunk 11 above)
Comment (Favna) [reply packages/in17n/src/register.ts:20]: FYI for @kyranet : other 2 duplicate it between `index.ts` and `register.ts`. I have a hunch that just register actually works now though with `/register.d.ts` just re-exporting `dist/register.d.ts` but I'd rather not risk the biscuit.

**Review by Favna — CHANGES_REQUESTED** [review]

File: packages/in17n/package.json [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+		"@types/i18next-fs-backend": "^1.0.0",
+		"discord.js": "^12.5.1"
+	},
+	"peerDependenciesMeta": {
+		"@types/i18next-fs-backend": {
+			"optional": true
+		},
+		"discord.js": {
+			"optional": true
+		}
```
Comment (Favna) [inline packages/in17n/package.json:38]:
```suggestion
		}
```

`src/lib/extensions/In17nMessage.ts` imports DJS stuff (and it cannot be a type import due to `Structures`). If you don't have DJS installed then TS will complain about that import, given you don't provide `--skipLibCheck` (which one shouldn't anyway)

**Review by Favna — CHANGES_REQUESTED** [review]

File: packages/i18next/src/register-discordjs.ts [hunk 26]
```diff
@@ -0,0 +1,134 @@
+export * from './register';
```
Comment (Favna) [inline packages/i18next/src/register-discordjs.ts:1]: Does this automatically cover these lines? I'm not sure...

```ts
SapphireClient.plugins.registerPostInitializationHook(I18nextPlugin[preGenericsInitialization], 'I18next-PreGenericsInitialization');
SapphireClient.plugins.registerPreLoginHook(I18nextPlugin[preLogin], 'I18next-PreLogin');
```

File: packages/i18next/tests/I18nextHandler.test.ts [hunk 27]
```diff
@@ -0,0 +1,22 @@
+import { isClass } from '@sapphire/utilities';
+import { I18nextHandler } from '../src';
+function structureTest(i18n: I18nextHandler) {
```
Comment (Favna) [inline packages/i18next/tests/I18nextHandler.test.ts:3]:
```suggestion
import { I18nextHandler } from '../src';

function structureTest(i18n: I18nextHandler) {
```

**Review by Favna — CHANGES_REQUESTED** [review]

File: .github/workflows/cd-in17n.yml [hunk 28]
```diff
@@ -0,0 +1,71 @@
+name: Continuous Delivery - in17n
```
Comment (Favna) [inline .github/workflows/cd-in17n.yml:1]:
```suggestion
name: Continuous Delivery - i18next
```

File: .github/workflows/cd-in17n.yml [hunk 29]
```diff
@@ -0,0 +1,71 @@
...
+
+on:
+  push:
+    branches:
+      - main
+    paths:
+      - packages/in17n/**
```
Comment (Favna) [inline .github/workflows/cd-in17n.yml:8]:
```suggestion
      - packages/i18next/**
```

File: .github/workflows/cd-in17n.yml [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+      - .all-contributorsrc
+      - LICENSE.md
+      - package.json
+      - .github/workflows/cd-in17n.yml
```
Comment (Favna) [inline .github/workflows/cd-in17n.yml:12]:
```suggestion
      - .github/workflows/cd-i18next.yml
```

File: .github/workflows/cd-in17n.yml [hunk 31]
```diff
@@ -0,0 +1,71 @@
...
+
+jobs:
+  BuildIn17n:
```
Comment (Favna) [inline .github/workflows/cd-in17n.yml:15]:
```suggestion
  BuildI18next:
```

File: .github/workflows/cd-in17n.yml [hunk 32]
```diff
@@ -0,0 +1,71 @@
...
+    name: Publish In17n
```
Comment (Favna) [inline .github/workflows/cd-in17n.yml:16]:
```suggestion
    name: Publish i18next
```

File: .github/workflows/cd-in17n.yml [hunk 33]
```diff
@@ -0,0 +1,71 @@
...
+          node-version: 14
+      - name: Restore CI Cache
+        uses: actions/cache@v1
+        with:
+          path: node_modules
+          key: ${{ runner.os }}-14-${{ hashFiles('**/yarn.lock') }}
+      - name: Install Dependencies
+        run: yarn --ignore-scripts --frozen-lockfile
+      - name: Build Code
+        run: yarn workspace @sapphire/plugin-in17n build
```
Comment (Favna) [inline .github/workflows/cd-in17n.yml:34]:
```suggestion
        run: yarn workspace @sapphire/plugin-i18next build
```

File: .github/workflows/cd-in17n.yml [hunk 34]
```diff
@@ -0,0 +1,71 @@
...
+          echo -e "\n# Initialize some useful variables"
+          REPO="https://${GITHUB_ACTOR}:${GITHUB_TOKEN}@github.com/${GITHUB_REPOSITORY}.git"
+          BRANCH_OR_TAG=`awk -F/ '{print $2}' <<< $GITHUB_REF`
+          CURRENT_BRANCH=`awk -F/ '{print $NF}' <<< $GITHUB_REF`
+          if [ "$BRANCH_OR_TAG" == "heads" ]; then
+            SOURCE_TYPE="branch"
+          else
+            SOURCE_TYPE="tag"
+          fi
+          echo -e "\n# Move to the in17n directory"
```
Comment (Favna) [inline .github/workflows/cd-in17n.yml:47]:
```suggestion
          echo -e "\n# Move to the i18next directory"
```

File: .github/workflows/cd-in17n.yml [hunk 35]
```diff
@@ -0,0 +1,71 @@
...
+          cd packages/in17n
```
Comment (Favna) [inline .github/workflows/cd-in17n.yml:48]:
```suggestion
          cd packages/i18next
```

File: .github/workflows/cd-in17n.yml [hunk 36]
```diff
@@ -0,0 +1,71 @@
...
+          echo -e "\n# Checkout the repo in the target branch"
+          TARGET_BRANCH="build-in17n"
```
Comment (Favna) [inline .github/workflows/cd-in17n.yml:50]:
```suggestion
          TARGET_BRANCH="build-i18next"
```

File: .github/workflows/cd-in17n.yml [hunk 37]
```diff
@@ -0,0 +1,71 @@
...
+          rsync -vaI ../../LICENSE.md out/
+          rsync -vaI README.md out/
+          rsync -vaI dist/ out/dist
+          rsync -vaI register* out/
+          echo -e "\n# Commit and push"
+          cd out
+          git add --all .
+          git config user.name "${GITHUB_ACTOR}"
+          git config user.email "${GITHUB_EMAIL}"
+          git commit -m "build: in17n build for ${GITHUB_SHA}" || true
```
Comment (Favna) [inline .github/workflows/cd-in17n.yml:66]:
```suggestion
          git commit -m "build: i18next build for ${GITHUB_SHA}" || true
```

**Review by Favna — CHANGES_REQUESTED** [review]
I also don't know if the file actually has no whitelines in the shell script part or if that's just GitHub, if it doesn't then please match the other files.

Also rename the file itself to `cd-i18next.yml`

File: packages/i18next/src/register-discordjs.ts [hunk 38]
```diff
@@ -0,0 +1,134 @@
...
+		 *   const guild = await driver.getRepository(GuildEntity).findOne({ id: message.guild.id });
+		 *   return guild?.language ?? 'en-US';
+		 * };
+		 * ```
+		 */
+		fetchLanguage: (message: any) => Promise<string | null> | string | null;
+	}
+
+	export interface ClientOptions {
+		i18n?: I18nOptions;
```
Comment (Favna) [inline packages/i18next/src/register-discordjs.ts:125]: Some documentation here would be nice (also requested by Kyra when Q was doing the PR for this line)

File: packages/i18next/src/lib/I18nextHandler.ts [hunk 39]
```diff
@@ -0,0 +1,146 @@
...
+	/**
+	 * Intitialises the handler by loading in the namespaces, passing the data to i18next, and filling in the {@link I18nextHandler#languages}.
+	 * @since 1.0.0
+	 */
+	public async init() {
+		const { namespaces, languages } = await this.walkLanguageDirectory(this.languagesDir);
+
+		i18next.use(Backend);
+		await i18next.init(
+			mergeDefault(
```
Comment (Favna) [inline packages/i18next/src/lib/I18nextHandler.ts:58]: See earlier comment by Kyra: https://github.com/sapphire-project/plugins/pull/29#discussion_r549511026

File: packages/i18next/src/lib/I18nextHandler.ts [hunk 40]
```diff
@@ -0,0 +1,146 @@
...
+	/**
+	 * @description Skips any files that don't end with `.json`.
+	 * @param dir The directory that should be walked.
+	 * @param namespaces The currently known namespaces.
+	 * @param folderName The currently walked folder.
+	 * @since 1.0.0
+	 * @protected
+	 */
+	protected async walkLanguageDirectory(dir: string, namespaces: string[] = [], folderName = '') {
+		const files = await readdir(dir);
```
Comment (Favna) [inline packages/i18next/src/lib/I18nextHandler.ts:129]: See earlier comment from Kyra: https://github.com/sapphire-project/plugins/pull/29#discussion_r549511604

**Review by Favna — CHANGES_REQUESTED** [review]

File: packages/i18next/src/lib/I18nextHandler.ts (hunk 39 above)
Comment (Nytelife26) [reply packages/i18next/src/lib/I18nextHandler.ts:58]: No mergeDefault will create issues, as we concluded on Discord. Until further notice this is resolved.

File: .github/CODEOWNERS [hunk 41]
```diff
@@ -1 +1,3 @@
-/packages/api @favna @QuantumlyTangled
+* @favna
```
Comment (Favna) [inline .github/CODEOWNERS:1]: ?

File: .github/workflows/cd-i18next.yml [hunk 42]
```diff
@@ -0,0 +1,69 @@
+name: Continuous Delivery - i18next
+on:
```
Comment (Favna) [inline .github/workflows/cd-i18next.yml:2]:
```suggestion
name: Continuous Delivery - i18next

on:
```

And check whitelines compared to other cd files.

**Review by Favna — CHANGES_REQUESTED** [review]

File: .github/CODEOWNERS (hunk 41 above)
Comment (Nytelife26) [reply .github/CODEOWNERS:1]: it was like that when I inherited the branch, haven't touched it.

Comment (Nytelife26) [reply .github/CODEOWNERS:1]: image
Can confirm, blame reveals it was commits prior to mine a whole 8 days ago that made that change. I have no idea.

Comment (Favna) [reply .github/CODEOWNERS]: Weird, well thanks for fixing it.

**Review by Favna — APPROVED** [review]

**Pull request merged**